File: ga/4omini_result_2024_7/result/1968D_mini/14_1968D.c
use std::io;
use std::cmp::{max, min};
use std::str::FromStr;

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();
    
    for _ in 0..t {
        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        let params: Vec<usize> = input.split_whitespace()
                                       .map(|x| x.parse().unwrap())
                                       .collect();
        let n = params[0];
        let k = params[1];
        let mut a = params[2] - 1; // Converts to zero-indexed
        let mut b = params[3] - 1;

        let m = min(n, k);
        let mut arr = vec![0; n];
        let mut aa = vec![0; n];

        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        for (i, val) in input.split_whitespace().enumerate() {
            aa[i] = val.parse().unwrap();
        }

        input.clear();
        io::stdin().read_line(&mut input).unwrap();
        for (i, val) in input.split_whitespace().enumerate() {
            arr[i] = val.parse().unwrap();
        }

        let mut p1 = 0_i64;
        let mut p2 = 0_i64;
        let mut p1_ = 0_i64;
        let mut p2_ = 0_i64;

        for x in 0..m {
            p1 = max(p1, aa[a] * (k - x as i64 + 1) + p1_);
            p1_ += aa[a];
            a = arr[a] - 1; // Converts to zero-indexed
            
            p2 = max(p2, aa[b] * (k - x as i64 + 1) + p2_);
            p2_ += aa[b];
            b = arr[b] - 1; // Converts to zero-indexed
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}

File: ga/gpt3_5_result_2024_6/result/1968D_gpt/14_1968D.rs
use std::cmp;

const N: usize = 200000;

fn min(a: i32, b: i32) -> i32 {
    if a < b { a } else { b }
}

fn maxl(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn main() {
    let mut t;
    let mut x;
    let mut n;
    let mut k;
    let mut a;
    let mut b;
    let mut a_;
    let mut b_;
    let mut m;
    let mut arr: [i32; N + 1] = [0; N + 1];
    let mut aa: [i32; N + 1] = [0; N + 1];

    let mut p1;
    let mut p2;
    let mut p1_;
    let mut p2_;

    let mut input_line = String::new();
    std::io::stdin().read_line(&mut input_line).unwrap();
    t = input_line.trim().parse().unwrap();

    while t > 0 {
        p1 = 0;
        p2 = 0;
        p1_ = 0;
        p2_ = 0;
        let mut i = -1;
        let mut i_ = -1;
        let mut j = -1;
        let mut j_ = -1;

        input_line.clear();
        std::io::stdin().read_line(&mut input_line).unwrap();
        let input_str: Vec<i32> = input_line
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();
        n = input_str[0];
        k = input_str[1];
        a = input_str[2];
        b = input_str[3];
        m = cmp::min(n, k);
        a_ = a;
        b_ = b;

        input_line.clear();
        std::io::stdin().read_line(&mut input_line).unwrap();
        let arr_input: Vec<i32> = input_line
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();
        for i in 1..=n {
            arr[i] = arr_input[i - 1];
        }

        input_line.clear();
        std::io::stdin().read_line(&mut input_line).unwrap();
        let aa_input: Vec<i32> = input_line
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();
        for i in 1..=n {
            aa[i] = aa_input[i - 1];
        }

        x = 1;
        while x <= m {
            p1 = maxl(p1, 1 * aa[a as usize] * (k - x + 1) + p1_);
            p1_ += aa[a as usize];
            a = arr[a as usize];
            p2 = maxl(p2, 1 * aa[b as usize] * (k - x + 1) + p2_);
            p2_ += aa[b as usize];
            b = arr[b as usize];

            x += 1;
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }

        t -= 1;
    }
}